File: backEnd/Microservices/AImodel/src/main/java/com/esprit/aimodel/service/ProviderStatusService.java
package com.esprit.aimodel.service;

import com.esprit.aimodel.dto.AiLiveStatus;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.List;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
public class ProviderStatusService {

    private final RestClient ollamaTagsRestClient;
    private final ObjectMapper objectMapper;
    private final String configuredModel;

    public ProviderStatusService(
            @Qualifier("ollamaTagsRestClient") RestClient ollamaTagsRestClient,
            ObjectMapper objectMapper,
            @Value("${spring.ai.ollama.chat.options.model:}") String configuredModel) {
        this.ollamaTagsRestClient = ollamaTagsRestClient;
        this.objectMapper = objectMapper;
        this.configuredModel = configuredModel != null ? configuredModel : "";
    }

    public AiLiveStatus liveStatus() {
        try {
            String body = ollamaTagsRestClient.get().uri("/api/tags").retrieve().body(String.class);
            List<String> names = parseOllamaModelNames(body, objectMapper);
            boolean modelReady = modelIsListed(names, configuredModel);
            return new AiLiveStatus("aimodel", "UP", true, configuredModel, modelReady);
        } catch (Exception e) {
            return new AiLiveStatus("aimodel", "UP", false, configuredModel, false);
        }
    }

    static List<String> parseOllamaModelNames(String json, ObjectMapper objectMapper) {
        List<String> names = new ArrayList<>();
        if (json == null || json.isBlank()) {
            return names;
        }
        try {
            JsonNode root = objectMapper.readTree(json);
            JsonNode models = root.get("models");
            if (models == null || !models.isArray()) {
                return names;
            }
            for (JsonNode model : models) {
                JsonNode nameNode = model.get("name");
                if (nameNode == null || !nameNode.isTextual()) {
                    continue;
                }
                String value = nameNode.asText().trim();
                if (!value.isEmpty()) {
                    names.add(value);
                }
            }
        } catch (Exception ignored) {
            // treat as no models
        }
        return names;
    }

    static boolean modelIsListed(List<String> names, String wanted) {
        if (wanted == null || wanted.isBlank() || names.isEmpty()) {
            return false;
        }
        if (names.contains(wanted)) {
            return true;
        }
        int colon = wanted.indexOf(':');
        String base = colon > 0 ? wanted.substring(0, colon) : wanted;
        for (String n : names) {
            if (base.equals(n) || (n != null && n.startsWith(base + ":"))) {
                return true;
            }
        }
        return false;
    }
}

File: backEnd/Microservices/aimodel-node/src/ai/providerStatusService.ts
import axios from "axios";
import { config, getRequestTimeoutMs } from "../config.js";
import type { AiLiveStatus } from "../dto.js";

function parseOllamaModelNames(payload: unknown): string[] {
  if (!payload || typeof payload !== "object") {
    return [];
  }

  const models = (payload as { models?: unknown }).models;
  if (!Array.isArray(models)) {
    return [];
  }

  return models
    .map((model) => {
      if (!model || typeof model !== "object") {
        return undefined;
      }
      const name = (model as { name?: unknown }).name;
      return typeof name === "string" ? name.trim() : undefined;
    })
    .filter((name): name is string => Boolean(name));
}

function modelIsListed(names: string[], wanted: string): boolean {
  if (!wanted.trim() || names.length === 0) {
    return false;
  }

  if (names.includes(wanted)) {
    return true;
  }

  const base = wanted.includes(":") ? wanted.slice(0, wanted.indexOf(":")) : wanted;
  return names.some((name) => name === base || name.startsWith(`${base}:`));
}

export async function liveStatus(): Promise<AiLiveStatus> {
  try {
    const response = await axios.get(`${config.ollamaBaseUrl}/api/tags`, {
      timeout: getRequestTimeoutMs(),
      headers: { Accept: "application/json" },
    });

    const names = parseOllamaModelNames(response.data);
    return {
      service: config.serviceName,
      status: "UP",
      ollamaReachable: true,
      model: config.ollamaModel,
      modelReady: modelIsListed(names, config.ollamaModel),
    };
  } catch {
    return {
      service: config.serviceName,
      status: "UP",
      ollamaReachable: false,
      model: config.ollamaModel,
      modelReady: false,
    };
  }
}
